### PR TITLE
feat: unify hierarchical patch merging and experiment recipes

### PR DESCRIPTION
**Pull request by Dafidofff**
## Summary

Add hierarchical patch merging to main, combining the implementation from #122 with the complementary local spatial-grid implementation. This PR supersedes #122 and targets main directly.

Instead of keeping one spatial resolution throughout a classifier, the new networks can repeatedly merge neighbouring 2×2 patches: for example, a 224×224 image becomes a 56×56 patch grid and then progresses through 28×28, 14×14 and 7×7 stages, typically doubling channel width at each transition. This supports hierarchical Hyena experiments and comparisons with flat architectures.

### Shared merging operation with two interfaces

`nvsubquadratic/modules/patch_merging.py` adds:

- `PatchMerging`: flat token input, optional prepended register row, configurable output width and normalization. Requires even spatial grids. Register tokens take a separate learned projection, preserving their conditioning role as patch resolution changes.
- `PatchMerging2D`: channels-last spatial grids, LayerNorm and C → 2C projection, with zero padding for odd heights/widths.
- `PatchEmbedHierarchical`: stride-four convolution and LayerNorm for the spatial hierarchy's input stem.

Both merger interfaces share the same 2×2 concatenation order. Tests verify identical outputs and gradients with compatible shared weights.

### Two hierarchical classifiers

- `ViT5HierarchicalClassificationNet`, from #122: configurable stage widths/depths, absolute position embeddings, and either pure patch tokens or register-row FiLM conditioning. Global average pooling excludes the register row.
- `ViT5HierarchicalNet`, recovered locally: four spatial-grid stages, normalized convolutional stem, no position/register tokens, configurable stochastic-depth schedule and odd-grid padding.

These are distinct experiment architectures; their checkpoints are not interchangeable. Existing parameter names within each implementation are preserved.

### Experiment recipes and data adapters

Includes the original `v6_hierarchical` ImageNet/CIFAR recipes, recovered `v5_patchmerge` spatial hierarchies and block-diagonal kernel variant, local ImageNet comparisons, and CIFAR hierarchical-versus-flat experiments.

CIFAR data recipes remain explicit: `cifar10.py` uses torchvision, while `cifar10_hf.py` retains the local Hugging Face source and augmentation settings. This avoids silently changing the meaning of either experiment family.

### Correctness and compatibility fixes

- Match register-row zero padding to projected-token dtype/device under autocast, avoiding unintended float32 promotion; add bf16 regression coverage.
- Correct the register-layout documentation identified in #122's review.
- Remove obsolete `Hyena(use_rope=False)` arguments from recovered configs.
- Count spatial hierarchy resolutions using ceil-halving for odd grids and validate required stage-width doubling.
- Update the documentation tracker and add a [usage/reconciliation guide](https://github.com/Dafidofff/nvSubquadratic/blob/codex/patch-merging-unified/examples/patch_merging/README.md).

## Review corrections

- Match flat/hierarchical Hyena optimizer-step budgets under gradient accumulation; make CIFAR label smoothing explicit and configurable. Remove unsupported claims of matched compute or a controlled merging-only ablation.
- Follow ViT5ClassificationNet's explicit no-decay normalization policy. Build spatial hierarchies lazily after seeding, preserving parameter metadata and serializable architecture settings. Keep direct model-building helpers for compatibility.
- Validate token hierarchy grids, merger dimensions and register layouts before forward; reject odd intermediate grids while allowing odd final grids.
- Start a one-block linear drop-path schedule at zero and enforce stage counts under optimized Python.
- Document and test filtering `network.head` when replacing the spatial classifier, preserving existing checkpoint keys. Restore the networks section in the test README.

## Scope and review order

The main-relative diff contains 43 files, primarily recovered source, experiment configs and tests. Review the merger module first, then the two networks, then the recipes/data adapters and their tests.

The branch includes current main and preserves #122's feature history. There is no dependency on merging #122 first. Local benchmark scripts already exist on main with identical Python ASTs and are not duplicated. Unrelated spatial-recall Gaussian changes, global FFT/mask edits, generated figures and historical result trackers are outside this PR.

## Validation

- 173 CPU tests passed across merger modules, both hierarchical networks, CIFAR adapters, optimizer groups and checkpoint/resume utilities.
- Regression coverage includes seeded lazy initialization, architecture serialization and CLI overrides, optimizer decay groups, explicit smoothing targets, the silent register-row mismatch case, odd-grid boundaries, one-block schedules, validation under `python -O`, and classifier replacement with ordinary/compiled checkpoint keys.
- Required pre-commit checks passed on all changed files, including lint, formatting, secret detection and license headers.
- Existing spatial checkpoints load strictly into the lazy-construction implementation and produce exactly equal logits for both scalar and block-diagonal kernel variants.
- GitHub checks are running on the updated head; their live status is shown in the PR checks panel.

Local validation used Python 3.12 / Torch 2.9.0+cu128 on CPU. Real Hyena CPU bf16 encounters an existing FFT shortcut-dtype assertion, so real-mixer CPU tests use float32; bf16 merging and hierarchy wiring are tested independently. ImageNet lazy-recipe tests substitute the unrelated Apex/DALI training-base dependency. No new training runs or accuracy claims are included.

### CODE REVIEW

File: examples/vit5_imagenet/v6_hierarchical/cifar10/hyena_flat.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+KERNEL_NUM_BLOCKS = 8
+KERNEL_OMEGA_0_MIN = 1.0
+KERNEL_OMEGA_0_MAX = 12.0
+KERNEL_HIDDEN_OMEGA_0 = 1.0
+KERNEL_OFF_BLOCK_SCALE = 0.1
+
+# ─── Training (identical to hier) ─────────────────────────────────────────────
+TARGET_EPOCHS = 100
+CIFAR10_TRAIN_SIZE = 50_000
+ITERS_PER_EPOCH = CIFAR10_TRAIN_SIZE // BATCH_SIZE
```
Comment (farhadrgh) [inline examples/vit5_imagenet/v6_hierarchical/cifar10/hyena_flat.py:94]: **The flat baseline trains ~4x longer than the hierarchical run it's compared against.**

`ITERS_PER_EPOCH` divides by the micro-batch (`BATCH_SIZE=64`) rather than the effective batch (`BATCH_SIZE * ACCUM_STEPS = 256`):

- `ITERS_PER_EPOCH = 50000 // 64 = 781` -> `TOTAL_ITERATIONS = 78_100`
- `trainer.py:208` sets `max_steps=cfg.train.iterations`, which Lightning counts as **optimizer** steps
- so the run consumes `78_100 * 256` = ~20M images = **400 data epochs**

`hyena_hier.py` (`BATCH_SIZE=256`, `accum=1`) gets `19_500` steps = **100 epochs**. The cosine schedule and 5-epoch warmup are stretched 4x as well.

That makes the module docstring's "same epoch budget, only the hierarchical structure differs" false, and the hier-vs-flat comparison invalid.

Fix: `ITERS_PER_EPOCH = CIFAR10_TRAIN_SIZE // (BATCH_SIZE * ACCUM_STEPS)` — which `v6_hierarchical/_base_config.py:123` already does correctly.

**Review by farhadrgh — COMMENTED** [review]
Review pass on the unified patch-merging recipes. The structural work looks solid — 69 new tests pass, and every leaf config builds.

My main concern is that the hierarchical-vs-flat/isotropic comparison this PR is built to enable has **four independent confounds** as currently configured, flagged inline below. Each one alone would make a hier-vs-flat result hard to interpret; together I don't think the runs would be comparable.

Separately (not inline, happy to expand): `stage_grid_sides` is floor-halved and never cross-checked against the supplied `patch_merge_cfgs`, so a mismatched merger silently corrupts the `register_row` GAP boundary instead of raising — and `initial_grid=6, num_stages=3` computes `[6, 3, 1]` for a 3x3 stage that can't be 2x2-merged at all. Also: a single-block net gets the *maximum* drop-path rate instead of 0 under the linear schedule; the stage-count `assert` vanishes under `python -O` and then `IndexError`s in `forward` via the hardcoded `i < 3`; and `out_proj` being a read-only property alias means `DropKeysFromCheckpoint(prefixes=('network.out_proj',))` matches zero keys, since the state_dict keys are `head.*`.

File: nvsubquadratic/modules/patch_merging.py [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+
+    Args:
+        dim: Input channel dimension C.
+    """
+
+    def __init__(self, dim: int):
+        """Initialize LayerNorm and the channel-doubling projection."""
+        super().__init__()
+        self.dim = dim
+        self.norm = nn.LayerNorm(4 * dim)
```
Comment (farhadrgh) [inline nvsubquadratic/modules/patch_merging.py:265]: **Weight decay is applied to every LayerNorm gain and the classifier bias in the hierarchical path, but not in the isotropic control.**

`PatchMerging` tags its norm explicitly (line 93):
```python
for p in self.norm.parameters():
    p._no_weight_decay = True
```
but `PatchMerging2D.norm` (here), `PatchEmbedHierarchical.norm` + `proj.bias` (line 204), and `ViT5HierarchicalNet.out_norm` / `head.bias` (`vit5_hierarchical.py:159-160`) do not.

`base_lightning_wrapper.py:104` then applies `default_weight_decay=0.05` to all of them. Running `_build_param_groups` on a small `build_hierarchical_net()` emits **24 warnings**, including `downsamplers.{0,1,2}.norm.{weight,bias}`, `out_norm.{weight,bias}`, `head.bias`, `stem.norm.*`.

The isotropic control (`ViT5ClassificationNet` + RMSNorm, bias-free head) produces **none** of these — so the ablation differs in regularization, not only architecture.

File: experiments/datamodules/cifar10.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        self.num_workers = num_workers
+        self.pin_memory = pin_memory
+        self.image_size = image_size
+        self.num_classes = num_classes
+        self.output_channels = num_classes
+
+        # timm Mixup/CutMix (only active during training)
+        use_mix = mixup > 0 or cutmix > 0
+        self._mixup_fn = (
+            Mixup(
```
Comment (farhadrgh) [inline experiments/datamodules/cifar10.py:103]: **Mixup silently inherits timm's `label_smoothing=0.1`.**

`Mixup` is constructed without `label_smoothing`, and timm 1.0.24 defaults it to `0.1` (verified). `CIFAR10DataModule` exposes only `mixup`/`cutmix`, so there's no way to configure or disable it.

Every v6_hierarchical CIFAR recipe (`hier_p*`, `flat_p*`, `hyena_flat`, `hyena_hier` — all `mixup=0.8 cutmix=1.0`) therefore trains with 0.1 label smoothing that appears nowhere in the config or the docstring.

The sibling `cifar10_hf.py:157` passes `label_smoothing` explicitly, and the ImageNet v6 recipe uses `smoothing=0.0` — so the two CIFAR families are silently training against different label targets.

File: examples/patch_merging/cifar10/hierarchical_hyena.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+from examples.patch_merging.cifar10._base import NUM_CLASSES, get_base_config
+from examples.vit5_imagenet.v5_patchmerge._base_config import build_hierarchical_net
+from experiments.default_cfg import ExperimentConfig
+
+
+def get_config() -> ExperimentConfig:
+    """Build hierarchical Hyena config for CIFAR-10 single-GPU training."""
+    config = get_base_config()
+    config.net = build_hierarchical_net(base_dim=80, fft_backend="torch_fft", num_classes=NUM_CLASSES)
```
Comment (farhadrgh) [inline examples/patch_merging/cifar10/hierarchical_hyena.py:50]: **`config.net` is assigned an already-instantiated `nn.Module` instead of a `LazyConfig`.**

`build_hierarchical_net()` returns a live `ViT5HierarchicalNet`. Three consequences, all verified against the runtime:

1. **Seeding doesn't apply.** `run.py:150` calls `load_config_from_file` (which runs `get_config()`, hence all weight init) *before* `pl.seed_everything(config.seed)` at `run.py:160`. So `config.seed` has no effect on initialization and runs aren't reproducible.
2. **The model is duplicated.** `instantiate(config.net)` hits `lazy_config.py:362` `copy.deepcopy` and returns a clone, so a second full copy (28.5M params at `base_dim=96`) stays alive in `config` for the whole run.
3. **No architecture reaches W&B.** `config_to_dict(config)` at `run.py:298` collapses `net` to the bare string `'nvsubquadratic.networks.vit5_hierarchical.ViT5HierarchicalNet'`, so the run records zero architecture info — and `--net.*` CLI overrides become impossible.

Same pattern in `v5_patchmerge/hierarchical_full_hyena.py:49`, `hierarchical_full_hyena_blockdiag.py`, and `local_comparison/hierarchical_hyena.py:46`. Every other config in the repo passes a `LazyConfig`.

**Review by farhadrgh — APPROVED** [review]
